core/src/main/java/org/infinispan/statetransfer/StateTransferLockImpl.java

```diff
@@ -122,36 +139,36 @@ public void notifyTopologyInstalled(int topologyId) {
          log.tracef("Signalling topology %d is installed", topologyId);
       }
       this.topologyId = topologyId;
-
       topologyLock.lock();
+      CompletableFuture<Void> oldFuture = topologyFuture;
       try {
-         topologyCondition.signalAll();
+         topologyFuture = new CompletableFuture();
       } finally {
          topologyLock.unlock();
+         remoteExecutor.execute(() -> oldFuture.complete(null));
       }
    }
 
    @Override
-   public void waitForTopology(int expectedTopologyId, long timeout, TimeUnit unit) throws InterruptedException {
+   public CompletableFuture<Void> topologyFuture(int expectedTopologyId) {
       if (topologyId >= expectedTopologyId)
-         return;
+         return null;
 
       if (trace) {
          log.tracef("Waiting for topology %d to be installed, current topology is %d", expectedTopologyId, topologyId);
       }
       topologyLock.lock();
       try {
-         long timeoutNanos = unit.toNanos(timeout);
-         while (topologyId < expectedTopologyId && timeoutNanos > 0) {
-            timeoutNanos = topologyCondition.awaitNanos(timeoutNanos);
+         if (topologyId >= expectedTopologyId) {
+            return null;
+         } else if (topologyId == expectedTopologyId - 1){
```

Comment:
Adding this check means our tests will rarely (if ever) go through the `thenCompose` branch, so you'll need a new test to exercise that branch. OTOH having to wait is already a pretty rare occurrence, so I'd rather remove this check instead.